spec/v1.0-draft/resource_descriptor.md

```diff
@@ -0,0 +1,139 @@
+# ResourceDescriptor field type specification
+
+Version v1.0-draft
+
+A size-efficient description of any mutable or immutable software artifact
+or resource.
+
+## Schema
+
+```json
+{
+  "name": "<NAME>",
+  "uri": "<RESOURCE URI>",
+  "digest": { "<ALGORITHM>": "<HEX VALUE>", ... },
+  "representation": { /* object */ },
+  "downloadLocation": "<RESOURCE URI>"
+  "mediaType": "<MIME TYPE>",
+  "annotations": { /* object */ }
+}
+```
+
+## Fields
+
+Though all fields are optional, a ResourceDescriptor MUST specify at least
+one of `uri` or `digest` at a minimum.
+
+`name` _string, optional_
+
+> Human-readable identifier to distinguish the resource or artifact locally.
+> The semantics are up to the producer and consumer.
+
+`uri` _[ResourceURI], optional_
+
+> A URI used to identify the resource or artifact. When a `digest` cannot be
+> computed for the resource or artifact, the producer SHOULD set this field.
+
+`digest` _[DigestSet], optional_
+
+> A set of crypographic digests of the contents of the resource or artifact.
+> When known, the producer SHOULD set this field to denote an immutable
+> artifact or resource. The producer and consumer must agree on acceptable
+> algorithms.
+
+`representation` _object, optional_
+
+> A representation of the contents of the resource or artifact.
+> The producer MAY use this field in scenarios where including the contents
+> of the resource/artifact directly in the attestation is deemed more
+> efficient for consumers than providing a reference to another location.
+>
+> The producer and consumer must agree on the semantics and acceptable
```

Comment:
I wonder if we have any more guidance (or just a disclaimer!) for how they're supposed to agree on this?  is it via the predicateType?  The identity of producer?

One consumer might have two producers creating the same predicate type but doing different things with representation or annotations.

Maybe it's a non-issue?

Reply:
Good point. This is pretty tricky, but agree that it would be useful to add more guidance round this. I don't think it will make sense in every case to map the `representation` to the upper-level `predicateType`. Per our discussion in https://github.com/slsa-framework/slsa/issues/649#issuecomment-1454161037 I'm thinking that the `uri and/or `mediaType` could be used as hints by the producer, which is directly related to the contents of `representation`.

Reply:
Yeah that makes a ton of sense of representation.

Would it make any sense for annotations to suggest following the mime convention of using `vnd.` prefixes for non standardized annotations?  (At the moment we have no standardized annotations but maybe some will come along)

Reply:
Suggested language

> There are no standardized annotations yet.  Users SHOULD follow [RFC 6838](https://www.rfc-editor.org/rfc/rfc6838.html#section-3.2) convention of prefixing annotation names with `x.`, `prs.`, or `vnd.` to avoid collisions with other users.

Reply:
This seems to be implying that annotations more broadly should be self-describing in some way. I don't know if a MIME type will always make sense. In that case, we might even just say that annotations SHOULD include either a TypeURI OR mime type, as you described.